Firestore/core/src/firebase/firestore/immutable/array_sorted_map.h

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_ARRAY_SORTED_MAP_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_ARRAY_SORTED_MAP_H_
+
+#include <assert.h>
+
+#include <algorithm>
+#include <array>
+#include <functional>
+#include <memory>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/immutable/map_entry.h"
+
+namespace firebase {
+namespace firestore {
+namespace immutable {
+
+/**
+ * A base class for implementing ArraySortedMap, which contains constants and
+ * types that don't depend upon the type of ArraySortedMap's type parameters.
+ */
+class ArraySortedMapBase {
+ public:
+  /**
+   * The type of size(). Note that this is not size_t specifically to save
+   * space in the TreeSortedMap implementation.
+   */
+  typedef uint32_t size_type;
+
+  /**
+   * The maximum size of an ArraySortedMap.
+   *
+   * This is the size threshold where we use a tree backed sorted map instead
+   * of an array backed sorted map. This is a more or less arbitrary chosen
+   * value, that was chosen to be large enough to fit most of object kind of
+   * Firebase data, but small enough to not notice degradation in performance
+   * for inserting and lookups. Feel free to empirically determine this
+   * constant, but don't expect much gain in real world performance.
+   */
+  static constexpr size_type kFixedSize = 25;
+};
+
+/**
+ * A thin wrapper around std::array that also maintains a dynamic size.
+ *
+ * ArraySortedMap does not actually contain its array: it contains a
+ * shared_ptr to a FixedArray
+ *
+ * @tparam T The contents of an element in the array.
+ * @tparam fixed_size the fixed size to use in creating the FixedArray.
+ */
+template <typename T, ArraySortedMapBase::size_type fixed_size>
+class FixedArray {
+ public:
+  typedef ArraySortedMapBase::size_type size_type;
+  typedef std::array<T, fixed_size> array_type;
+  typedef typename array_type::iterator iterator;
+  typedef typename array_type::const_iterator const_iterator;
+
+  /**
+   * Appends to this array, copying from the given src_begin up to but not
+   * including the src_end.
+   */
+  template <typename SourceIterator>
+  void append(SourceIterator src_begin, SourceIterator src_end) {
+    auto copy_end = std::copy(src_begin, src_end, end());
+    size_ = static_cast<size_type>(copy_end - begin());
```

Comment:
I think it would be better to move checks that current size doesn't exceed max size from Map to FixedArray. The invariant of this class (on top of the invariant of std::array) seems to be "size_ reflects the portion of the buffer used". However, the class allows violating this invariant by exceeding capacity without any error reporting. Enforcing checks on the class using FixedArray makes it easier to miss a case (this class is very small, and all places where size_ is modified are very clear and close together).

Reply:
As it stands, FixedArray doesn't seem to really be tied to anything specific to ArraySortedMap, so I think it makes sense for it to be an independent class.